Warn when operating system end of life is approaching

MarkEWaite opened the pull request:
## Warn when operating system end of life is approaching

Jenkins controllers run on many different operating systems.  The [Linux Support Policy](https://www.jenkins.io/doc/administration/requirements/linux/) states that the Jenkins project does not support Jenkins on operating systems that are no longer supported by the operating system provider.  Administrative monitors added in this pull request will alert administrators of the end of life of the operating system that is running Jenkins so that they can upgrade to a supported operating system.

The operating system end of life dates are taken from the [endoflife.date site](https://endoflife.date/).

### Early end of life for RHEL 7, CentOS 7, and others

This pull request proposes to accelerate the end of life date for Red Hat Enterprise Linux 7, CentOS Linux 7, Scientific Linux 7, and Oracle Linux 7.  Those operating systems will be unsupported by their providers after June 30, 2024.  This pull request proposes to end Jenkins support for those operating systems by December 31, 2023.

Those operating systems are already unsupported by the `systemd` rpm based installer for Jenkins.  The [CentOS 7 container image](https://hub.docker.com/_/centos) that is the basis for the Jenkins `centos7` container image is already deprecated.  It was last updated Nov 16, 2020.

Command line git 1.8 as included with those operating systems is far behind the command line git 2.x versions on every other operating system tested by the Jenkins project.  It lacks features that are available on every other operating system supported by Jenkins.

OpenSSH 7.4 as included with those operating systems is far behind the OpenSSH versions on every other operating system tested by the Jenkins project.  It lacks features that are available on every other operating system supported by Jenkins.

If it will be more acceptable, I'm willing to split this pull request into two pull requests, one that adds the Admin monitors for scheduled operating system end of life (like Ubuntu 18.04, Alpine Linux 3.14, etc.) and one that proposes to accelerate the Red Hat Enterprise Linux 7 end of life.

I'm also willing to submit a Jenkins Enhancement Proposal for the accelerated end of life, if there is a strong preference for that to be discussed in a JEP.

### Screenshots

The end of life admin monitor for CentOS 7 looks like this when displayed to the administrator from the main page:

![main-page-with-pop-up-expanded](https://github.com/jenkinsci/jenkins/assets/156685/f7af2049-15dd-43f8-8000-db68789c97ef)

The administrative monitor looks like this from the "Manage Jenkins" page

![manage-jenkins-with-warning-visible](https://github.com/jenkinsci/jenkins/assets/156685/ae875461-20b6-459e-a72b-c3934d0a426c)


The administrative monitor covers the operating systems that will reach end of life by the end of 2024 and beyond.

![manage-jenkins-admin-screen](https://github.com/jenkinsci/jenkins/assets/156685/0fcf659e-acfb-40c1-ac92-eef5b5ddc121)

### Testing done

Tested interactively with a private build of the Jenkins `centos7` container image to confirm that the admin monitor displays correctly when run with CentOS 7 and that it is not displayed when running on other operating systems like Red Hat Enterprise Linux 8.

Tested interactively that the administrative monitor checkboxes in the "Manage Jenkins" / "System" page correctly enable and disable individual monitors.

Confirmed the published end of life dates match the dates in the source files, with the exception of the RHEL 7 accelerated end of life date.

Automated tests include sample data from each of the target operating systems.  Input / output exceptions reading the operating system data files are the only lines not covered by automated tests.

### Proposed changelog entries

- Warn administrators when their Linux operating system is approaching end of life
- Announce early end of life for Red Hat Enterprise Linux 7 and its derivatives

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] The Jira issue, if it exists, is well-described.
- [x] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)).
  - Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [x] There is automated testing or an explanation as to why this change has no tests.
- [x] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [x] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [x] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [x] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [x] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

* @gounthar from the Platform SIG
* @kmartens27 from the Docs SIG
* @NotMyFault and @krisstern because they attended a Docs Office Hours session and discussed the idea with me
* @timja because I used the Java revision monitor as the basis for this code and placed him as one of the copyright holders on the source files that I copied to start the pull request.

### Maintainer checklist

Before the changes are marked as `ready-for-merge`:

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).


<a href="https://gitpod.io/#https://github.com/jenkinsci/jenkins/pull/7913"><img src="https://gitpod.io/button/open-in-gitpod.svg"/></a>

MarkEWaite committed:
Ubuntu 18.04 end of life admin monitor

First prototype that includes a user interface and localized messages
with parameters to support the specific operating system.

MarkEWaite committed:
Use a base class for common features

Admin monitors are disabled and enabled by a user interface action.
There is not an obvious way to detect that a date has passed so that it
should be enabled again.

MarkEWaite committed:
Better comment

MarkEWaite committed:
Make the logging clearer

MarkEWaite committed:
Split pattern check to separate method

MarkEWaite committed:
Add a message to show after end of life

MarkEWaite committed:
Suppress most logging

MarkEWaite committed:
Minor logging improvements

MarkEWaite committed:
Use actual end of life date for Ubuntu 18.04

MarkEWaite committed:
Logger not needed

MarkEWaite committed:
Add Alpine 3.14 end of life admin monitor

MarkEWaite committed:
Add Alpine 3.15 end of life monitor

MarkEWaite committed:
Add Alpine 3.16 end of life monitor

MarkEWaite committed:
Add Alpine 3.17 end of life monitor

MarkEWaite committed:
Spotless cleanup

MarkEWaite committed:
Add Fedora 37, 38, and Red Hat 7

MarkEWaite committed:
More accurate regex to detect version

MarkEWaite committed:
Add Debian 10 end of life admin monitor

MarkEWaite committed:
Add a few tests

MarkEWaite committed:
Add another test

MarkEWaite committed:
Fix checkstyle violations

MarkEWaite committed:
Fix spotbugs warning

MarkEWaite committed:
Correct the copyright

MarkEWaite committed:
Make test viable on non-Linux machines

Do not depend on the existence of /etc/os-release

MarkEWaite committed:
Name the setup function more accurately

MarkEWaite committed:
Test a file with non-matching content

MarkEWaite committed:
Test the Alpine end of life admin monitors

MarkEWaite committed:
Add more tests

MarkEWaite committed:
Test Fedora 37 admin monitor

MarkEWaite committed:
Add RHEL 7 test

MarkEWaite committed:
Announce RHEL 7 early end of support immediately

MarkEWaite committed:
Use consistent logging calls

MarkEWaite committed:
Add Red Hat 8 to RHEL 7 as a test

MarkEWaite committed:
Add Red Hat 8 to RHEL 7 as a test

MarkEWaite committed:
Use past date to report RHEL 7 end of life

MarkEWaite committed:
Remove Red Hat 8 from RHEL 7 end of life

MarkEWaite committed:
Log more information

MarkEWaite committed:
Display end of support date

Incorrect to declare end of support as the date when we begin warning
the user about end of life.

MarkEWaite committed:
Add descriptions of the admin monitors

MarkEWaite committed:
Use a better redirect URL

MarkEWaite committed:
Group admin monitors better by display name

MarkEWaite committed:
Remove unnused message.properties files

MarkEWaite committed:
Add description.jelly file

MarkEWaite committed:
Better description of the admin monitors

MarkEWaite committed:
Add endOfSupportDate getter

Needed for UI

MarkEWaite committed:
Fix minor indentation error

MarkEWaite committed:
Each monitor must be independently disabled

MarkEWaite committed:
Repurpose identifier for use in disabled property

Don't need the idenitfier anywhere except in the initial constructor
call that is later used to enable and disable the admin monitor.

MarkEWaite committed:
Add os-release tests and data files

MarkEWaite committed:
Remove unused URL arg to constructor

MarkEWaite committed:
Reduce duplication in data

Easier to test with symbolic constants that are consistently named across
all the classes.

Includes tests of os-release files from each of the end of life operating systems

MarkEWaite committed:
Fix inaccurate comment

MarkEWaite committed:
Use constant for documentation URL

MarkEWaite committed:
Remove RHEL 8 from RHEL 8 Admin Monitor

MarkEWaite committed:
Fedora 36 security support ends 18 May 2023

MarkEWaite committed:
Fedora 37 support ends 15 Dec 2023

timja commented on core/src/main/resources/jenkins/monitor/EndOfLifeAdminMonitor/message.jelly:
```diff
@@ -0,0 +1,45 @@
+        <form id="${it.id}" method="post" action="${rootURL}/${it.url}/act" name="${it.id}">
+            <f:submit name="yes" value="${%More Info}"/>
+            <l:isAdmin>
+                <f:submit value="${%Ignore} ${it.displayName}" name="no"/>
```
The UI from the screenshots looks quite broken, buttons should not be that long and the heading shouldn't span multiple lines like that..
this may not be enough to fix it but it's a start
```suggestion
                <f:submit value="${%Dismiss}" name="no"/>
```

MarkEWaite replied:
Thanks.  Replaced long text with "Ignore" to be consistent with other admin monitors.  See 8ae1172bc1fead511aa33a5b8be2261e2ccbc654

MarkEWaite replied:
I could also remove the "and its derivatives" from the heading of the dialog box and rely on the administrator to read the body text in order to realize that their operating system is also end of life, even though it is not named "Red Hat Enterprise Linux".

timja replied:
would it be better to put the actual OS, that way there's less user confusion and the heading is a lot simpler / shorter ? 

e.g. 

```
➜  azure-vm-agents-plugin git:(key-based-auth) ✗ docker run -it ubuntu bash
root@622f1c9a01ce:/# cat /etc/os-release | grep ^NAME=
NAME="Ubuntu"
root@622f1c9a01ce:/#
exit
➜  azure-vm-agents-plugin git:(key-based-auth) ✗ docker run -it centos:7 bash
Unable to find image 'centos:7' locally
7: Pulling from library/centos
6717b8ec66cd: Pull complete
Digest: sha256:be65f488b7764ad3638f236b7b515b3678369a5124c47b8d32916d6487418ea4
Status: Downloaded newer image for centos:7
[root@f369fa8f3f9c /]# cat /etc/os-release | grep ^NAME=
NAME="CentOS Linux"
```

I had a look around and there's not a better way that I can see, Java `os.name` property only shows linux

MarkEWaite replied:
> I had a look around and there's not a better way that I can see, Java `os.name` property only shows linux

The contents of `/etc/os-release` that match the version number specification provide a good phrasing for the detected operating system.  In the test data it looks like this:

```
os-release-alpine-3.14:PRETTY_NAME="Alpine Linux v3.14"
os-release-alpine-3.15:PRETTY_NAME="Alpine Linux v3.15"
os-release-alpine-3.16:PRETTY_NAME="Alpine Linux v3.16"
os-release-alpine-3.17:PRETTY_NAME="Alpine Linux v3.17"
os-release-centos-7:PRETTY_NAME="CentOS Linux 7 (Core)"
os-release-debian-10:PRETTY_NAME="Debian GNU/Linux 10 (buster)"
os-release-fedora-36:PRETTY_NAME="Fedora Linux 36 (Container Image)"
os-release-fedora-37:PRETTY_NAME="Fedora Linux 37 (Container Image)"
os-release-oracle-7:PRETTY_NAME="Oracle Linux Server 7.9"
os-release-redhat-7:PRETTY_NAME="Red Hat Enterprise Linux Server 7.9 (Maipo)"
os-release-scientific-7:PRETTY_NAME="Scientific Linux 7.9 (Nitrogen)"
os-release-ubuntu-18.04:PRETTY_NAME="Ubuntu 18.04.6 LTS"
```

The definition of `/etc/os-release` states that version information in `PRETTY_NAME` is optional, but in all the cases noted here, it includes a version number.  All the test data files tracked in the platform labeler plugin also include a version number in the `PRETTY_NAME` value (except openSUSE Tumbleweed that is defined to not have a version number).

timja commented on core/src/main/resources/jenkins/monitor/EndOfLifeAlpine315AdminMonitor/description.properties:
```diff
@@ -0,0 +1 @@
+blurb = Warns administrators when Alpine 3.15 is approachinig end of life and will be unsupported
```
typo throughout
```suggestion
blurb=Warns administrators when Alpine 3.15 is approaching end of life and will be unsupported
```

MarkEWaite replied:
Thanks.  Fixed in 1753bee8b73a5735b2a2b1f74a1f1a627a5aa9f0 .  The spaces around the `=` sign exist in 93 other locations that use `blurb`.  1753bee8b73a5735b2a2b1f74a1f1a627a5aa9f0 did not change any of those locations, assuming that they are harmless.

timja commented on core/src/main/java/jenkins/monitor/EndOfLifeAdminMonitor.java:
```diff
@@ -0,0 +1,179 @@
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+/* Package protected so that it won't be visible outside this package */
+class EndOfLifeAdminMonitor extends AdministrativeMonitor {
```
I wonder if this should be just one admin monitor for Operating System End of Life.

you could still have it as an extension point so the logic could be separated but you wouldn't have all these pointless admin monitors in the enabled list as a controller is only ever going to run on one OS.

<img width="529" alt="image" src="https://user-images.githubusercontent.com/21194782/235467421-d64f6053-1b46-4a33-ae34-975e5b783812.png">

MarkEWaite replied:
I like that suggestion very much.  I'll need to investigate further

timja commented on core/src/main/java/jenkins/monitor/EndOfLifeAdminMonitor.java:
```diff
@@ -0,0 +1,179 @@
+        boolean matched = false;
+        LOGGER.log(FINE, "Reading file {1} with monitor {0}", new Object[] {identifier, dataFile});
+        if (dataFile.isFile()) {
+            try (BufferedReader reader = new BufferedReader(Files.newBufferedReader(dataFile.toPath(), Charset.defaultCharset()))) {
```
can we use utf-8 here instead of the default charset?

MarkEWaite replied:
Done in 89c39fac725aa05af5f46d6cb262d3d5e64cf1a3

timja commented on core/src/main/java/jenkins/monitor/EndOfLifeAlpine314AdminMonitor.java:
```diff
@@ -0,0 +1,101 @@
+
+@Extension
+@Restricted(NoExternalUse.class)
+@Symbol("endOfLifeAlpine314AdminMonitor")
```
symbols should not contain their extension point

```suggestion
@Symbol("endOfLifeAlpine314")
```

timja reviewed (commented):
Some feedback from looking at the code and screenshots.

In terms of reviewing this it would be easier if the mechanism and one or two operating systems were added and then follow ups added for the rest.

MarkEWaite committed:
Fix spelling error in description

MarkEWaite committed:
Read data file with UTF-8 character set

MarkEWaite committed:
Do not use extension point name in symbol

MarkEWaite committed:
Use single word 'Ignore' in button, not name of monitor

kmartens27 reviewed (approved):
Looks good to me, thanks for your work on this @MarkEWaite

timja commented on core/src/main/resources/jenkins/monitor/EndOfLifeAdminMonitor/message.jelly:
```diff
@@ -0,0 +1,45 @@
+                <f:submit value="${%Ignore}" name="no"/>
+            </l:isAdmin>
+        </form>
+        <h2>${it.displayName}</h2>
```
might be worth making this smaller with e.g.
(not pulled locally currently)
```suggestion
        <h2 class="h3">${it.displayName}</h2>
```

MarkEWaite replied:
I like the smaller look that generates.  Added as 545315e523e013cff71f1b4c598b03d198e64b7d

MarkEWaite committed:
Use a smaller layout for monitor name

MarkEWaite commented:
@timja I think that it makes sense for me to attempt a reimplementation as a single administrative monitor that detects the operating system and version and then uses that to decide if the administrator should be warned.  Some portions of this implementation should be usable, but more as a learning exercise for me than as specific code.

gounthar reviewed (approved):
Thanks a lot for this implementation, Mark. 🙏 
When displaying the message related to the end of life, would there be a way to show a link to jenkins.io to the end-user so he/she can choose a more recent version (yes, we would have to create and maintain such a page), or would this be opening a can of worms?

MarkEWaite committed:
Include operating system version in RHEL 7 check

Includes a TODO comment to remove the RHEL 8 entry that I use for
testing and development

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite commented:
I've converted this to draft because I believe that the concept is correct but this implementation needs to be discarded and replaced with a single administrative monitor that checks the operating system on the controller.

MarkEWaite committed:
Fix regex syntax error

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite committed:
First prototype using a data file

MarkEWaite committed:
Start the conversion to a single monitor

MarkEWaite committed:
Latest experiments with end of life data as JSON

MarkEWaite committed:
More exploring and evolving

MarkEWaite committed:
Remove Alpine admin monitor

MarkEWaite committed:
Remove Fedora admin monitor

MarkEWaite committed:
Remove Debian admin monitor

MarkEWaite committed:
Remove Ubuntu admin monitor

MarkEWaite committed:
Throw IOException rather than check for null

IOException should be uncommon and mistakes should be detected by
automated tests

MarkEWaite committed:
Stop inheriting from EndOfLifeAdminMonitor

MarkEWaite committed:
Simplify initialization logic

Else clauses not required due to break.

MarkEWaite committed:
Add isActivated implementation

MarkEWaite committed:
Add setDisabled test method

MarkEWaite committed:
Minor formatting changes

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite committed:
Test some additional methods

MarkEWaite committed:
Call super in constructor

MarkEWaite committed:
Minor updates from study of upstream class

MarkEWaite committed:
Format JSON data

MarkEWaite committed:
Fix formatting errors

MarkEWaite committed:
Fix assertion

MarkEWaite committed:
Better variable naming

MarkEWaite committed:
Explain the purpose of ignoreEndOfLife flag

MarkEWaite committed:
Remove unused import

MarkEWaite committed:
Store the pretty name, not the pattern

MarkEWaite committed:
Merge remote-tracking branch 'upstream/master' into file-based-admin-monitor

MarkEWaite committed:
Test pretty name with sample os-release data files

MarkEWaite committed:
Remove RedHat 7 specific monitor

MarkEWaite committed:
Set RHEL 8 end of life early for test

MarkEWaite committed:
Ongoing debugging

MarkEWaite committed:
Describe the monitor

MarkEWaite committed:
Honor start date

MarkEWaite committed:
Add pretty name as monitor attribute

MarkEWaite committed:
Add end of support date

MarkEWaite committed:
Remove earlier implementations

MarkEWaite committed:
Move end of life data

MarkEWaite committed:
Rename end of support to end of life

End of life is used throughout the monitor

MarkEWaite committed:
Improve message with after end of life date status

MarkEWaite committed:
Undo .gitignore change

MarkEWaite committed:
Rename effective date to end of life date

MarkEWaite committed:
Rename pretty name to operating system name

MarkEWaite committed:
More rename of prettyName

MarkEWaite committed:
Fix logging mistakes

MarkEWaite committed:
Add RHEL 8

MarkEWaite committed:
Default start date is 6 months before end of life

MarkEWaite committed:
Add operating system end of life patterns and dates

MarkEWaite committed:
Fix JSON file formatting

MarkEWaite committed:
Simplify logging code

MarkEWaite committed:
Better logging

MarkEWaite committed:
Add Alma and Rocky Linux tests and patterns

MarkEWaite committed:
Add methods to help test automation

MarkEWaite committed:
Remove unused methods

MarkEWaite committed:
Start warning RHEL 7 users immediately

MarkEWaite committed:
Suppress spotbugs traversal_in warning - false positive

MarkEWaite committed:
Throw IOException on bad data file

Fail tests quickly when a a bad data file is used

MarkEWaite committed:
Better exception message

MarkEWaite committed:
Correct RHEL 8 end of life date

MarkEWaite committed:
Enable isActivated test

MarkEWaite committed:
Increase test coverage

MarkEWaite commented:
The pull request has been changed to use a single admin monitor to check for many different operating systems.  Interactive testing was performed as described in "Testing done" with the most recent version of these changes.  Automated tests have been created to cover most of the branches and statements in the new class.

The Red Hat Enterprise Linux 7 end of life stated in this pull request is intended to happen at about the same time as Jenkins 2.425.1, the 2nd LTS baseline after 2.401.1.  That gives weekly users 6 months of warning and gives LTS users 3 months of warning.

NotMyFault commented on core/src/test/java/jenkins/monitor/EndOfLifeOperatingSystemAdminMonitorTest.java:
```diff
@@ -0,0 +1,223 @@
+
+    @Test
+    public void testGetOperatingSystemName() {
+        /* Operating system name depends on the operating system runnning test */
```
```suggestion
        /* Operating system name depends on the operating system running test */
```

NotMyFault commented on core/src/test/resources/jenkins/monitor/os-release-centos-7:
```diff
@@ -0,0 +1,16 @@
+CENTOS_MANTISBT_PROJECT_VERSION="7"
+REDHAT_SUPPORT_PRODUCT="centos"
+REDHAT_SUPPORT_PRODUCT_VERSION="7"
+
```
```suggestion
```

MarkEWaite replied:
I would rather not make this change.

The file is copied from the [CentOS 7 container image](https://hub.docker.com/r/jenkins/jenkins/tags?page=1&name=centos7) that we use for the Jenkins controller image.  Removing the empty line will make the test data file different than the file on the actual operating system.  It would be harmless in this case, since the test does not depend on that blank line, but I think it is better to keep the test data aligned with the file as it exists on CentOS 7.

NotMyFault reviewed (approved):
Thanks!

MarkEWaite committed:
Fix spelling error

Co-authored-by: Alexander Brandes <brandes.alexander@web.de>

timja commented on core/src/main/java/jenkins/monitor/EndOfLifeOperatingSystemAdminMonitor.java:
```diff
@@ -0,0 +1,243 @@
+
+@Extension
+@Restricted(NoExternalUse.class)
+@Symbol("operatingSystemEndOfLife")
+public class EndOfLifeOperatingSystemAdminMonitor extends AdministrativeMonitor {
```
generally the symbol name and class name would roughly match. is there a reason it's worded differently?

MarkEWaite replied:
I think that the symbol name better describes the intent of the monitor.  I'm happy to rename the class from  `EndOfLifeOperatingSystemAdminMonitor` to `OperatingSystemEndOfLifeAdminMonitor`.  Would that address your concern or is there something more that I'm missing?

timja replied:
yes renaming the class would do it

MarkEWaite replied:
Renamed in cfb7be5306457da80e473c689c6228452a53ad09 

Thanks for the recommendation!

timja commented on core/src/main/resources/jenkins/monitor/EndOfLifeOperatingSystemAdminMonitor/message.properties:
```diff
@@ -0,0 +1,9 @@
+End_of_Life_Heading={0} in Jenkins
+Upcoming_End_of_Life=<p class="jenkins-!-margin-bottom-0">You are running Jenkins on {0}.&nbsp;&nbsp;\
```
generally non breaking spaces aren't recommended as they can't be styled and can cause weird issues.
```suggestion
Upcoming_End_of_Life=<p class="jenkins-!-margin-bottom-0">You are running Jenkins on {0}. \
```

MarkEWaite replied:
Happy to remove the non-breaking spaces.  I didn't like the layout without the non-breaking spaces because the text seemed to remove all space between the full stop and the first word of the following sentence.  I'll remove it to address your concern and capture a new screen shot for further discussion.

timja commented on core/src/main/resources/jenkins/monitor/EndOfLifeOperatingSystemAdminMonitor/message.properties:
```diff
@@ -0,0 +1,9 @@
+Upcoming_End_of_Life=<p class="jenkins-!-margin-bottom-0">You are running Jenkins on {0}.&nbsp;&nbsp;\
+  Jenkins will no longer support {0} after <strong>{1}</strong>.\
+  <p>Please plan your upgrade to a supported operating system <strong>before {1}</strong>.</p>\
+  <p>Refer to <a href="https://www.jenkins.io/redirect/operating-system-end-of-life" target="_blank" rel="noopener noreferrer">the documentation</a> for details.</p>
```
should a query parameter include the OS maybe?

MarkEWaite replied:
Good suggestion.

I don't see any query parameter support in the redirect use on jenkins.io,. but I do see a very effective use of a fragment at the end of the URL.  When reporting a bug, we can append the Jira component number to the `report-issue` URL and it will pass that component number to the final destination

For example https://www.jenkins.io/participate/report-issue/redirect/#21710  becomes a link that will report a bug, enhancement, or security issues to the workflow-aggregator-plugin.  

I could do something similar here by adding some form of the operating system name and version as a fragment at the end of the redirect URL.  Initially it will likely be ignored, but in the future we could use the fragment for navigation within the page.

timja replied:
it would allow analytics to be done if it was a query parameter 😄

MarkEWaite replied:
Implemented in 24d991a6e5011ecc38b81da19f8b7d4fcb6d41f7

timja commented on core/src/main/resources/jenkins/monitor/EndOfLifeOperatingSystemAdminMonitor/message.properties:
```diff
@@ -0,0 +1,9 @@
+  Jenkins will no longer support {0} after <strong>{1}</strong>.\
+  <p>Please plan your upgrade to a supported operating system <strong>before {1}</strong>.</p>\
+  <p>Refer to <a href="https://www.jenkins.io/redirect/operating-system-end-of-life" target="_blank" rel="noopener noreferrer">the documentation</a> for details.</p>
+Past_End_of_Life=<p class="jenkins-!-margin-bottom-0">You are running Jenkins on {0}.&nbsp;&nbsp;\
```
```suggestion
Past_End_of_Life=<p class="jenkins-!-margin-bottom-0">You are running Jenkins on {0}. \
```

MarkEWaite committed:
Remove non-breaking spaces from message

Non-breaking spaces are not recommended.

Co-authored-by: Tim Jacomb <21194782+timja@users.noreply.github.com>

MarkEWaite committed:
Rename class to match symbol

https://github.com/jenkinsci/jenkins/pull/7913/files#r1199814150

MarkEWaite committed:
Sort end of life data alphabetically

MarkEWaite committed:
Included OS query parameter and fragment in docs URL

The fragment may be useful in the future for analytics.  Harmless to
send the fragment to www.jenkins.io and www.jenkins.io may be able to
make use of it to direct the reader to information that is specific to
their operating system.

MarkEWaite committed:
Scientific Linux 8 does not exist

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

timja commented on core/src/test/java/jenkins/monitor/OperatingSystemEndOfLifeAdminMonitorTest.java:
```diff
@@ -0,0 +1,334 @@
+    }
+
+    private String docsUrl(String component) {
+        return "https://www.jenkins.io/redirect/operating-system-end-of-life?q=" + component + "#" + component;
```
why are we using a query string and a fragment?

MarkEWaite replied:
Fragment can be used to redirect as is done with the "Submit an issue" redirect.  Query string is used in case there is a way to get metrics from it.

I'm fine any combination of them.  Let me know which you prefer and I'll make it that way.

The base URL is enough for my needs currently.  The base URL plus fragment is enough for the existing use cases of www.jenkins.io redirects.  The base URL plus fragment plus query string was a wild guess that it might be useful someday.

timja replied:
Why can fragment be used for redirect and query string can't?
i.e. doesn't query string do both analytics and the ability to redirect to a specific page later if we wish

MarkEWaite replied:
https://github.com/jenkins-infra/jenkins.io/blob/master/content/participate/report-issue/redirect.html.haml uses the fragment so that https://www.jenkins.io/participate/report-issue/redirect/#15543 opens a page that will submit https://issues.jenkins.io/secure/CreateIssueDetails!init.jspa?pid=10172&issuetype=1&priority=4&components=15543 so that the bug report includes the correct component.

The fragment is extracted from the URL in the JavaScript in that page so that it can be included in the URL submitted to Jira.  I wasn't sure if there is a way to do the same thing with a query string

timja replied:
Yes you can get it with: `new URLSearchParams(window.location.search).get('q')`

MarkEWaite replied:
Thanks.  I'll remove the fragment so that the code only passes the query parameter

MarkEWaite replied:
Removed in 0a35a4a1db5ddad106e01f63cf48109e7e5028b3

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite committed:
Remove fragment, only need query parameter

https://github.com/jenkinsci/jenkins/pull/7913#discussion_r1200991347
is the comment that prompted the improvement.

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

timja committed:
Merge branch 'master' into file-based-admin-monitor

MarkEWaite committed:
Merge branch 'master' into file-based-admin-monitor

NotMyFault committed:
Merge branch 'master' into file-based-admin-monitor
